procedures/statement_performance_analyzer.sql

```diff
@@ -0,0 +1,792 @@
+  `SUM_SELECT_SCAN` bigint(20) unsigned NOT NULL,
+  `SUM_SORT_MERGE_PASSES` bigint(20) unsigned NOT NULL,
+  `SUM_SORT_RANGE` bigint(20) unsigned NOT NULL,
+  `SUM_SORT_ROWS` bigint(20) unsigned NOT NULL,
+  `SUM_SORT_SCAN` bigint(20) unsigned NOT NULL,
+  `SUM_NO_INDEX_USED` bigint(20) unsigned NOT NULL,
+  `SUM_NO_GOOD_INDEX_USED` bigint(20) unsigned NOT NULL,
+  `FIRST_SEEN` timestamp NULL DEFAULT NULL,
+  `LAST_SEEN` timestamp NULL DEFAULT NULL,
+  INDEX (SCHEMA_NAME, DIGEST)
+) DEFAULT CHARSET=utf8';
+
+    -- Do the action
+    -- The actions snapshot, ... requires a fresh snapshot - create it now
+    IF (v_force_new_snapshot
+           OR in_action = 'snapshot'
+           OR (in_action = 'overall' AND in_table IS NULL)
+           OR (in_action = 'save' AND v_tmp_digests_table_exists <> 'TEMPORARY')
+       ) THEN
+        IF (v_tmp_digests_table_exists = 'TEMPORARY') THEN
+            IF (@sys.debug = 'ON') THEN
+                SELECT 'DROP TEMPORARY TABLE IF EXISTS tmp_digests' AS 'Debug';
+            END IF;
+            DROP TEMPORARY TABLE IF EXISTS tmp_digests;
+        END IF;
+        SET @sys.tmp.SQL = REPLACE(REPLACE(v_digest_table_template, '%{TEMPORARY}', 'TEMPORARY '), '%{TABLE_NAME}', 'tmp_digests');
+        IF (@sys.debug = 'ON') THEN
+            SELECT @sys.tmp.SQL AS 'Debug';
+        END IF;
+        PREPARE stmt_create_table FROM @sys.tmp.SQL;
+        EXECUTE stmt_create_table;
+        DEALLOCATE PREPARE stmt_create_table;
+
+        SET @sys.tmp.SQL = CONCAT('INSERT INTO tmp_digests SELECT * FROM ',
+                                  IF(in_table IS NULL OR in_action = 'save', 'performance_schema.events_statements_summary_by_digest', v_quoted_table));
+        IF (@sys.debug = 'ON') THEN
+            SELECT @sys.tmp.SQL AS 'Debug';
+        END IF;
+        PREPARE stmt_insert FROM @sys.tmp.SQL;
+        EXECUTE stmt_insert;
+        DEALLOCATE PREPARE stmt_insert;
+    END IF;
+
+    -- Go through the remaining actions
+    IF (in_action IN ('create_table', 'create_tmp')) THEN
+        IF (in_action = 'create_table') THEN
+            SET @sys.tmp.SQL = REPLACE(REPLACE(v_digest_table_template, '%{TEMPORARY}', ''), '%{TABLE_NAME}', v_quoted_table);
+        ELSE
+            SET @sys.tmp.SQL = REPLACE(REPLACE(v_digest_table_template, '%{TEMPORARY}', 'TEMPORARY '), '%{TABLE_NAME}', v_quoted_table);
+        END IF;
+        IF (@sys.debug = 'ON') THEN
+            SELECT @sys.tmp.SQL AS 'Debug';
+        END IF;
+        PREPARE stmt_create_table FROM @sys.tmp.SQL;
+        EXECUTE stmt_create_table;
+        DEALLOCATE PREPARE stmt_create_table;
+
+    ELSEIF (in_action = 'save') THEN
+        SET @sys.tmp.SQL = CONCAT('DELETE FROM ', v_quoted_table);
+        IF (@sys.debug = 'ON') THEN
+            SELECT @sys.tmp.SQL AS 'Debug';
+        END IF;
+        PREPARE stmt_delete FROM @sys.tmp.SQL;
+        EXECUTE stmt_delete;
+        DEALLOCATE PREPARE stmt_delete;
```

Comment:
This pattern is repeated over and over, it seems like this is one of those cases where we could extract this to something like an "execute_prepared_statement" procedure, and just pass in the SQL string to prepare. I would do something like:

CALL execute_prepared_stmt(CONCAT('DELETE FROM ', v_quoted_table));

And have the routine itself set the input string in to a user variable, check debug, prepare/execute/deallocate.